rts/lra/lra-test/lra-test-basic/src/test/java/io/narayana/lra/arquillian/resource/LRAParticipantWithStatusURI.java

```diff
@@ -0,0 +1,125 @@
+
+import org.eclipse.microprofile.lra.annotation.Compensate;
+import org.eclipse.microprofile.lra.annotation.Complete;
+import org.eclipse.microprofile.lra.annotation.Forget;
+import org.eclipse.microprofile.lra.annotation.ParticipantStatus;
+import org.eclipse.microprofile.lra.annotation.Status;
+import org.eclipse.microprofile.lra.annotation.ws.rs.LRA;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_CONTEXT_HEADER;
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_RECOVERY_HEADER;
+
+@ApplicationScoped
+@Path(LRAParticipantWithStatusURI.LRA_PARTICIPANT_PATH)
+public class LRAParticipantWithStatusURI {
+    public static final String LRA_PARTICIPANT_PATH = "participant-with-status-reporting";
+    public static final String TRANSACTIONAL_CLOSE_PATH = "close-work";
+    public static final String TRANSACTIONAL_CANCEL_PATH = "cancel-work";
+    public static final String FORGET_COUNT_PATH = "forget-count";
+
+    private static final AtomicInteger forgetCount = new AtomicInteger(0);
+    private static final AtomicInteger compensateCount = new AtomicInteger(0);
+    private static final AtomicInteger completeCount = new AtomicInteger(0);
+
+    @GET
+    @Path(TRANSACTIONAL_CLOSE_PATH)
+    @LRA
+    public Response closeLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.ok(lraId.toASCIIString()).build();
+    }
+
+    @GET
+    @Path(TRANSACTIONAL_CANCEL_PATH)
+    @LRA(cancelOn = Response.Status.INTERNAL_SERVER_ERROR)
+    public Response abortLRA(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(lraId.toASCIIString()).build();
+    }
+
+    @GET
+    @Path(FORGET_COUNT_PATH)
+    public int getForgetCount() {
+        return forgetCount.get();
+    }
+
+    @PUT
+    @Path("compensate")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Compensate
+    public Response compensateWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
+            throws NotFoundException {
+        compensateCount.incrementAndGet();
+
+        return Response.status(500).entity(ParticipantStatus.FailedToCompensate.name()).build();
+    }
+
+    @PUT
+    @Path("complete")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Complete
+    public Response completeWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
+            throws NotFoundException {
+        completeCount.incrementAndGet();
+
+        return Response.status(500).entity(ParticipantStatus.FailedToComplete.name()).build();
```

Comment:
Don't Narayana retry on 500 status code? In the spec we use 200 and only the response body (here correct FailedToComplete) to distinguish failed states.

Reply:
That is only if the method returns an invalid enum value. The actual meaning of the relevant codes is:

200 means the resource has completed (https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/Complete.java#L116)

500 means the resource has failed to complete (https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/Complete.java#L129)

If any other code is returned (or, in the 500 case, the body does not correspond to a valid state)  then the implementation SHOULD either keep retrying or attempt to discover the status (https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/Complete.java#L163)

Reply:
My mistake, thanks. This can be resolved.